### PR TITLE
Fakeiteasy 4.0.0

### LINKED ISSUE #27: Support FakeItEasy 4.0.0
Currently we're binary incompatible, as the signature of FakeItEasy's `Wrapping` options building method has changed.

Unless it's too tricky, we might as well continue supporting 3.x.

### PR DESCRIPTION
**Pull request by blairconrad**
Fixes #27.

### CODE REVIEW

File: tests/SelfInitializingFakes.Tests.Earliest/Acceptance/BinarySerialization.cs [hunk 1]
```diff
@@ -75,7 +75,11 @@ public interface IService
                 .x(() =>
                 {
                     int i;
+#if FAKEITEASY3
                     DateTime dt = new DateTime(2017, 1, 24);
+#else
+                    DateTime dt = DateTime.MinValue;
+#endif
```
Comment (thomaslevesque) [inline tests/SelfInitializingFakes.Tests.Earliest/Acceptance/BinarySerialization.cs:82]: I don't understand this. Why this change?

**Comment by blairconrad** [comment]
Forgot to adjust the package dependencies!

File: tests/SelfInitializingFakes.Tests.Latest/SelfInitializingFakes.Tests.Latest.csproj [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  <ItemGroup Condition=" '$(TargetFramework)' == 'net452' ">
+    <Reference Include="System" />
+    <Reference Include="Microsoft.CSharp" />
+  </ItemGroup>
+
+  <ItemGroup Condition=" '$(TargetFramework)' == 'netcoreapp1.0' ">
+    <Compile Remove="..\SelfInitializingFakes.Tests.Earliest\Acceptance\BinarySerialization.cs" />
+  </ItemGroup>
+
+</Project>
```
Comment (thomaslevesque) [inline tests/SelfInitializingFakes.Tests.Latest/SelfInitializingFakes.Tests.Latest.csproj:43]: Wouldn't it be easier to just put both projects in the same folder?

File: src/SelfInitializingFakes/SelfInitializingFake.of.T.cs [hunk 3]
```diff
@@ -64,6 +69,20 @@ internal SelfInitializingFake(Func<TService> serviceFactory, IRecordedCallReposi
...
+        {
+            return options =>
+            {
+#if FEATURE_NETCORE_REFLECTION
+                var optionsType = options.GetType().GetTypeInfo();
+#else
+                var optionsType = options.GetType();
+#endif
+                var wrapping = optionsType.GetMethod("Wrapping", new[] { typeof(TService) });
+                wrapping.Invoke(options, new object[] { wrappedService });
```
Comment (thomaslevesque) [inline src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: I realize that performance isn't really critical, but it wouldn't hurt to generate and cache a delegate ;)

File: tests/SelfInitializingFakes.Tests.Earliest/Acceptance/BinarySerialization.cs (hunk 1 above)
Comment (blairconrad) [reply tests/SelfInitializingFakes.Tests.Earliest/Acceptance/BinarySerialization.cs:82]: https://github.com/FakeItEasy/FakeItEasy/issues/1124. The test was relying on old behaviour.

File: src/SelfInitializingFakes/SelfInitializingFake.of.T.cs (hunk 3 above)
Comment (blairconrad) [reply src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: I was lazy. And figured it could be done later. Will do.

File: tests/SelfInitializingFakes.Tests.Latest/SelfInitializingFakes.Tests.Latest.csproj (hunk 2 above)
Comment (blairconrad) [reply tests/SelfInitializingFakes.Tests.Latest/SelfInitializingFakes.Tests.Latest.csproj:43]: I forget about this option. Maybe? The only difference in the csprojs would be one include line and changed paths on a couple of removes…

I worry about the pulled-in FakeItEasy.dll fighting, since we need different versions. Mightn't they collide?

**Comment by blairconrad** [comment]
Thanks for the review, @thomaslevesque! I appreciate you donating your time.

File: src/SelfInitializingFakes/SelfInitializingFake.of.T.cs (hunk 3 above)
Comment (blairconrad) [reply src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: Um. I don't see `Delegate.CreateDelegate` available in .NET Standard 1.6. Is it? Am I chasing the wrong approach?

File: tests/SelfInitializingFakes.Tests.Earliest/Acceptance/BinarySerialization.cs (hunk 1 above)
Comment (thomaslevesque) [reply tests/SelfInitializingFakes.Tests.Earliest/Acceptance/BinarySerialization.cs:82]: Ah, yes. Thanks for the reminder.

File: tests/SelfInitializingFakes.Tests.Latest/SelfInitializingFakes.Tests.Latest.csproj (hunk 2 above)
Comment (thomaslevesque) [reply tests/SelfInitializingFakes.Tests.Latest/SelfInitializingFakes.Tests.Latest.csproj:43]: Good point. In FakeItEasy.Analyzer I handled this by using different folders for `obj`, but it was ugly. Maybe your approach is better after all.

File: src/SelfInitializingFakes/SelfInitializingFake.of.T.cs (hunk 3 above)
Comment (thomaslevesque) [reply src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: Oh. Is `MethodInfo.CreateDelegate` available?

Comment (blairconrad) [reply src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: In .NET Standard 1.6, but not in .NET 4.0. I can `#if` it.

Comment (blairconrad) [reply src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: Wow. I am failing miserably at caching a delegate.

Comment (thomaslevesque) [reply src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: I can give it a try if you want, I've done similar things quite often.

Comment (blairconrad) [reply src/SelfInitializingFakes/SelfInitializingFake.of.T.cs:82]: As we discussed offline, I am not going to bother. Can always optimize later, right?
